kernel/kernel-api/src/test/scala/io/delta/kernel/internal/types/DataTypeJsonSerDeSuite.scala

```diff
@@ -245,6 +274,249 @@ object DataTypeJsonSerDeSuite {
+                new StringType("SPARK.UTF8_LCASE"), true), true),
+            new StructType()
+              .add("c1", new StringType("SPARK.UTF8_LCASE"), false)
+              .add("c2", new StringType("ICU.UNICODE"), true)
+              .add("c3", StringType.STRING), true))
+          .add("b2", LongType.LONG), true)
+        .add("a3", new ArrayType(
+          new MapType(
+            new StringType("ICU.UNICODE_CI"),
+            new StructType()
+              .add("b1", new StringType("SPARK.UTF8_LCASE"), false), false), false), true)
+    )
+  )
+
+  val SAMPLE_JSON_TO_TYPES_WITH_COLLATION_DIFFERING = Seq(
+    (
+      "\"string\"", new StringType("SPARK.UTF8_LCASE")
+    ),
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", "\"string\"", true,
+          metadataJson = Some(
+            s"""{"$COLLATIONS_METADATA_KEY" :
+               | {"a1" : "SPARK.UTF8_LCASE"}}""".stripMargin)))),
+      new StructType()
+        .add("a1", StringType.STRING, true)
+    ),
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", "\"string\"", true,
+          metadataJson = Some(
+            s"""{"$COLLATIONS_METADATA_KEY" :
+               | {"a1" : "SPARK.UTF8_LCASE"}}""".stripMargin)))),
+      new StructType()
+        .add("a1", new StringType("ICU.UNICODE"), true)
+    ),
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", structTypeJson(Seq(
+          structFieldJson("b1", "\"string\"", true,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b1" : "ICU.UNICODE"}}""".stripMargin)))), true),
+        structFieldJson("a2", structTypeJson(Seq(
+          structFieldJson("b1", arrayTypeJson("\"string\"", false), true,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b1.element" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+          structFieldJson("b2", mapTypeJson("\"string\"", "\"string\"", true), false,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b2.key" : "ICU.UNICODE_CI",
+                 |  "b2.value" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+          structFieldJson("b3", arrayTypeJson("\"string\"", false), true),
+          structFieldJson("b4", mapTypeJson("\"string\"", "\"string\"", false), false))), true),
+        structFieldJson("a3", structTypeJson(Seq(
+          structFieldJson("b1", "\"string\"", false),
+          structFieldJson("b2", arrayTypeJson("\"integer\"", false), true))), false,
+          metadataJson = Some(
+            s"""{"$COLLATIONS_METADATA_KEY"
+               | : {"b1" : "SPARK.UTF8_LCASE"}}""".stripMargin)))),
+      new StructType()
+        .add("a1", new StructType()
+          .add("b1", new StringType("ICU.UNICODE")), true)
+        .add("a2", new StructType()
+          // In json, this field has "SPARK.UTF8_LCASE" collation
+          .add("b1", new ArrayType(StringType.STRING, false))
```

Comment:
why do we need these negative tests? If I understand correctly, as long as the `StructType.equals` is implemented correctly, then this should work. May be just add one specific test where it is not matching?

Reply:
okay, make sense since we have StringTypeSuite where we test this